src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs

```diff
@@ -762,31 +790,282 @@ public void Dispose()
         protected virtual void Dispose(bool disposing)
         {
-            if (!this._disposed)
+            if (!_disposed)
             {
                 if (disposing)
                 {
-                    _sender.Dispose();
+                    _sender?.Dispose();
+                    _pingComplete?.Dispose();
                 }
 
                 _disposed = true;
             }
         }
 
-        // Count of pings sent per each trace route hop.
-        // Default = 3 (from Windows).
-        // If we change 'DefaultTraceRoutePingCount' we should change 'ConsoleTraceRouteReply' resource string.
-        private const int DefaultTraceRoutePingCount = 3;
+        // Uses the SendAsync() method to send pings, so that Ctrl+C can halt the request early if needed.
+        private PingReply SendCancellablePing(
+            IPAddress targetAddress,
+            int timeout,
+            byte[] buffer,
+            PingOptions pingOptions,
+            Stopwatch timer = null)
+        {
+            try
+            {
+                _sender = new Ping();
+                _sender.PingCompleted += OnPingComplete;
 
-        /// Create the default send buffer once and cache it.
-        private const int DefaultSendBufferSize = 32;
-        private static byte[] s_DefaultSendBuffer = null;
+                timer?.Start();
+                _sender.SendAsync(targetAddress, timeout, buffer, pingOptions, this);
+                _pingComplete.Wait();
+                timer?.Stop();
+                _pingComplete.Reset();
 
-        private bool _disposed;
+                if (_pingCompleteArgs.Cancelled)
+                {
+                    // The only cancellation we have implemented is on pipeline stops via StopProcessing().
+                    throw new PipelineStoppedException();
+                }
 
-        private readonly Ping _sender = new Ping();
+                if (_pingCompleteArgs.Error != null)
+                {
+                    throw new PingException(_pingCompleteArgs.Error.Message, _pingCompleteArgs.Error);
+                }
 
-        private const string TestConnectionExceptionId = "TestConnectionException";
+                return _pingCompleteArgs.Reply;
+            }
+            finally
+            {
+                _sender.Dispose();
+                _sender = null;
+            }
+        }
+
+        // This event is triggered when the ping is completed, and passes along the eventargs so that we know
+        // if the ping was cancelled, or an exception was thrown.
+        private static void OnPingComplete(object sender, PingCompletedEventArgs e)
+        {
+            ((TestConnectionCommand)e.UserState)._pingCompleteArgs = e;
+            ((TestConnectionCommand)e.UserState)._pingComplete.Set();
+        }
+
+        /// <summary>
+        /// The class contains information about the source, the destination and ping results.
+        /// </summary>
+        public class PingStatus
```

Comment:
Does this need to be public?

Reply:
Not really, it just seems weird to output to the user an object type that can't be at least referenced directly. Making it internal seems like it would hamper users from being able to handle it easily.

Reply:
You did make the constructors internal.  I'm fine with this.

Reply:
Is it worth moving them outside the main TestConnectionCommand class, do you think? It _does_ seem slightly weird that a public class has such a complicated type name :grin:

Reply:
There is an option to use `PSObject` wrapped `PingReply` if we have no need for high performance.
Do we really want to have PingStatus?

Reply:
I think we can provide more value with a custom type than with PSObject; I do not see a compelling reason to accept slower and messy ETS additions here. 🤷‍♂